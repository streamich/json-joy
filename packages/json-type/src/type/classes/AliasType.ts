import {printTree} from 'tree-dump/lib/printTree';
import type {Printable} from 'tree-dump/lib/types';
import type {Type} from '../../type';
import type {ModuleType} from './ModuleType';

export class AliasType<K extends string, T extends Type> implements Printable {
  public constructor(
    public readonly system: ModuleType,
    public readonly id: K,
    public readonly type: T,
  ) {}

  public getType(): Type {
    return this.type;
  }

  public resolve(): AliasType<string, Type> {
    return this.system.resolve(this.id);
  }

  public toString(tab: string = '') {
    return this.id + printTree(tab, [(tab) => this.type.toString(tab)]);
  }
}
