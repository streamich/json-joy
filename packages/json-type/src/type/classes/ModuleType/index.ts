import {printTree} from 'tree-dump/lib/printTree';
import {Walker} from '../../../schema/Walker';
import {TypeBuilder} from '../../TypeBuilder';
import {AliasType} from '../AliasType';
import type {Printable} from 'tree-dump/lib/types';
import type {KeySchema, ModuleSchema, ObjSchema, Schema, TypeMap} from '../../../schema';
import type {Type} from '../../../type';
import type {RefType} from '../RefType';

export class ModuleType implements Printable {
  public static readonly from = (module: ModuleSchema): ModuleType => {
    const type = new ModuleType();
    type.import(module);
    return type;
  };

  public readonly t = new TypeBuilder(this);

  public readonly aliases: Map<string, AliasType<string, any>> = new Map();

  /**
   * @todo Add ability fetch object of given type by its ID, analogous to
   * GraphQL "nodes".
   */
  public readonly alias = <K extends string, T extends Type>(id: K, type: T): AliasType<K, T> => {
    const existingAlias = this.aliases.get(id);
    if (existingAlias) return existingAlias as AliasType<K, T>;
    const alias = new AliasType<K, T>(this, id, type);
    this.aliases.set(id, alias);
    return alias;
  };

  public readonly unalias = <K extends string>(id: K): AliasType<K, Type> => {
    const alias = this.aliases.get(id);
    if (!alias) throw new Error(`Alias [id = ${id}] not found.`);
    return <AliasType<K, Type>>alias;
  };

  public readonly hasAlias = (id: string): boolean => this.aliases.has(id);

  public readonly resolve = <K extends string>(id: K): AliasType<K, Type> => {
    const alias = this.unalias(id);
    return alias.type.kind() === 'ref' ? this.resolve<K>((alias.type as RefType).ref() as K) : alias;
  };

  public exportTypes() {
    const result: Record<string, unknown> = {};
    for (const [id, alias] of this.aliases.entries()) {
      result[id] = alias.getType().getSchema();
    }
    return result;
  }

  public import(module: ModuleSchema): void {
    const map: TypeMap = {};
    for (const alias of module.keys) {
      map[alias.key] = alias.value as Schema;
    }
    const expandObjFields = (aliasOfObj: string | ObjSchema): KeySchema[] => {
      const obj = typeof aliasOfObj === 'string' ? (map[aliasOfObj] as ObjSchema) : aliasOfObj;
      if (!obj || obj.kind !== 'obj') throw new Error('NO_OBJ');
      if (obj.extends) {
        const uniqueFields: Map<string, KeySchema> = new Map();
        for (const parent of obj.extends) {
          const parentFields = expandObjFields(parent);
          for (const field of parentFields) uniqueFields.set(field.key, field);
        }
        delete obj.extends;
        for (const field of obj.keys) uniqueFields.set(field.key, field);
        obj.keys = [...uniqueFields.values()];
      }
      return obj.keys;
    };
    Walker.walk(module, {
      onType: (type) => {
        if (type.kind !== 'obj') return;
        if (type.extends) expandObjFields(type);
      },
    });
    this.importTypes(map);
  }

  public importTypes<Aliases extends TypeMap>(
    aliases: Aliases,
  ): {
    readonly [K in keyof Aliases]: AliasType<
      K extends string ? K : never,
      /** @todo Replace `any` by inferred type here. */ any
    >;
  } {
    const result = {} as any;
    for (const id in aliases) result[id] = this.alias(id, this.t.import(aliases[id]));
    return result;
  }

  public toString(tab: string = '') {
    return (
      'Module' +
      printTree(tab, [
        (tab) =>
          'aliases' +
          printTree(
            tab,
            [...this.aliases.values()].map((alias) => (tab) => alias.toString(tab)),
          ),
      ])
    );
  }
}
