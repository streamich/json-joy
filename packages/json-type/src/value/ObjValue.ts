import {ModuleType} from '../type/classes/ModuleType';
import {Value} from './Value';
import {FnValue} from './FnValue';
import type {Printable} from 'tree-dump/lib/types';
import type * as classes from '../type';
import type {TypeBuilder} from '../type/TypeBuilder';

export type UnObjType<T> = T extends classes.ObjType<infer U> ? U : never;
export type UnObjValue<T> = T extends ObjValue<infer U> ? U : never;
export type UnObjFieldTypeVal<T> = T extends classes.KeyType<any, infer U> ? U : never;
export type ObjFieldToTuple<F> = F extends classes.KeyType<infer K, infer V> ? [K, V] : never;
export type ToObject<T> = T extends [string, unknown][] ? {[K in T[number] as K[0]]: K[1]} : never;
export type ObjValueToTypeMap<F> = ToObject<{
  [K in keyof F]: ObjFieldToTuple<F[K]>;
}>;

export type Ensure<T, X> = T extends X ? T : X;

export class ObjValue<T extends classes.ObjType<any>> extends Value<T> implements Printable {
  public static new = (system: ModuleType = new ModuleType()) => new ObjValue({}, system.t.obj);

  public get system(): classes.ModuleType {
    return (this.type as T).getSystem();
  }

  public get t(): TypeBuilder {
    return this.system.t;
  }

  public keys(): string[] {
    const type = this.type as T;
    return type.keys.map((field: classes.KeyType<string, any>) => field.key);
  }

  public get<K extends keyof ObjValueToTypeMap<UnObjType<T>>>(
    key: K,
  ): Value<
    ObjValueToTypeMap<UnObjType<T>>[K] extends classes.Type ? ObjValueToTypeMap<UnObjType<T>>[K] : classes.Type
  > {
    const field = this.type!.getField(<string>key);
    if (!field) throw new Error('NO_FIELD');
    const data = (this.data as Record<string, unknown>)[<string>key];
    return new Value(data, field.val) as any;
  }

  public fn<K extends keyof ObjValueToTypeMap<UnObjType<T>>>(
    key: K,
  ): FnValue<
    Ensure<
      ObjValueToTypeMap<UnObjType<T>>[K] extends classes.Type ? ObjValueToTypeMap<UnObjType<T>>[K] : classes.Type,
      classes.FnType<any, any, any>
    >
  > {
    const val = this.get(key);
    return new FnValue(val.data, val.type as any);
  }

  public field<F extends classes.KeyType<any, any>>(
    field: F | ((t: TypeBuilder) => F),
    data: classes.ResolveType<UnObjFieldTypeVal<F>>,
  ): ObjValue<classes.ObjType<[...UnObjType<T>, F]>> {
    field = typeof field === 'function' ? field((this.type as classes.ObjType<any>).getSystem().t) : field;
    (this.data as any)[field.key] = data;
    const type = this.type!;
    const system = type.system;
    if (!system) throw new Error('NO_SYSTEM');
    type.keys.push(field);
    return this as any;
  }

  public add<K extends string, V extends classes.Type>(
    key: K,
    type: V | ((t: TypeBuilder) => V),
    data: classes.ResolveType<V>,
  ) {
    const system = (this.type as classes.ObjType<any>).getSystem();
    const t = system.t;
    type = typeof type === 'function' ? type(t) : type;
    return this.field(t.Key(key, type), data);
  }

  public set<K extends string, V extends classes.Type>(key: K, value: Value<V>) {
    return this.add(key, value.type!, value.data);
  }

  public merge<O extends ObjValue<any>>(obj: O): ObjValue<classes.ObjType<[...UnObjType<T>, ...UnObjType<O['type']>]>> {
    Object.assign(this.data as object, obj.data);
    const type = this.type!;
    const system = type.system;
    if (!system) throw new Error('NO_SYSTEM');
    type.keys.push(...type.keys);
    return this as any;
  }

  public name(): string {
    return 'ObjValue';
  }
}
