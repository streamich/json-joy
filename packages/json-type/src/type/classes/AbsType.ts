import type {Printable} from 'tree-dump/lib/types';
import type * as schema from '../../schema';
import type {SchemaExample} from '../../schema';
import {Value} from '../../value';
import type {BaseType, ModuleType} from '../types';

export abstract class AbsType<S extends schema.Schema> implements BaseType<S>, Printable {
  /** Default type system to use, if any. */
  public system?: ModuleType;

  public readonly _validators: [validator: (value: unknown) => void, name?: string][] = [];

  constructor(public readonly schema: S) {}

  public sys(system: ModuleType | undefined): this {
    this.system = system;
    return this;
  }

  public getSystem(): ModuleType {
    const system = this.system;
    if (!system) throw new Error('NO_SYSTEM');
    return system;
  }

  public kind(): S['kind'] {
    return this.schema.kind;
  }

  public value(data: schema.TypeOf<S>) {
    return new Value<this>(data as any, this);
  }

  /**
   * @todo Add ability to export the whole schema, including aliases.
   */
  public getSchema(): S {
    return this.schema;
  }

  /**
   * Sets a custom runtime validator for this type.
   *
   * @param validator Function that validates the value of this type.
   * @returns `this` for chaining.
   */
  public validator(validator: (value: schema.TypeOf<S>) => void, name?: string): this {
    this._validators.push([validator as any, name]);
    return this;
  }

  public options(options: schema.Optional<Omit<S, 'kind'>>): this {
    // biome-ignore lint: unused variables are intentional
    const {kind, ...sanitizedOptions} = options as any;
    Object.assign(this.schema, sanitizedOptions);
    return this;
  }

  public title(title: string): this {
    this.schema.title = title;
    return this;
  }

  public intro(intro: string): this {
    this.schema.intro = intro;
    return this;
  }

  public description(description: string): this {
    this.schema.description = description;
    return this;
  }

  public default(value: schema.Schema['default']): this {
    this.schema.default = value;
    return this;
  }

  public example(
    value: schema.TypeOf<S>,
    title?: SchemaExample['title'],
    options?: Omit<SchemaExample, 'value' | 'title'>,
  ): this {
    const examples = (this.schema.examples ??= []);
    const example: SchemaExample = {...options, value};
    if (typeof title === 'string') example.title = title;
    examples.push(example);
    return this;
  }

  public getOptions(): schema.Optional<S> {
    // biome-ignore lint: unused variables are intentional
    const {kind, ...options} = this.schema;
    return options as any;
  }

  public alias<K extends string>(name: K) {
    return this.getSystem().alias(name, this);
  }

  protected toStringTitle(): string {
    return this.kind();
  }

  protected toStringOptions(): string {
    const options = this.getOptions() as schema.Display;
    const title = options.title || options.intro || options.description;
    if (!title) return '';
    return JSON.stringify(title);
  }

  public toString(tab: string = ''): string {
    const options = this.toStringOptions();
    return this.toStringTitle() + (options ? ` ${options}` : '');
  }
}
