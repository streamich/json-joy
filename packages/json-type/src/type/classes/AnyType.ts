import type * as schema from '../../schema';
import {AbsType} from './AbsType';

export class AnyType extends AbsType<schema.AnySchema> {}
