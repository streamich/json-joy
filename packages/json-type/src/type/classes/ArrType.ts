import {printTree} from 'tree-dump';
import * as schema from '../../schema';
import type {SchemaOf, Type} from '../types';
import {AbsType} from './AbsType';
import type {TypeExportContext} from './ModuleType/TypeExportContext';

export class ArrType<
  T extends Type | void = any,
  const Head extends Type[] = any,
  const Tail extends Type[] = any,
> extends AbsType<
  schema.ArrSchema<
    T extends void ? schema.Schema : SchemaOf<T extends Type ? T : never>,
    {[K in keyof Head]: SchemaOf<Head[K]>},
    {[K in keyof Tail]: SchemaOf<Tail[K]>}
  >
> {
  constructor(
    public readonly _type?: T,
    public readonly _head?: Head,
    public readonly _tail?: Tail,
    options?: schema.Optional<schema.ArrSchema>,
  ) {
    super(schema.s.Array(schema.s.any, options) as schema.ArrSchema<any, any, any>);
  }

  public head<const H extends Type[]>(...head: H): ArrType<T, H, Tail> {
    (this as any)._head = head as any;
    return this as any;
  }

  public tail<const X extends Type[]>(...tail: X): ArrType<T, Head, X> {
    (this as any)._tail = tail as any;
    return this as any;
  }

  public min(min: schema.ArrSchema['min']): this {
    this.schema.min = min;
    return this;
  }

  public max(max: schema.ArrSchema['max']): this {
    this.schema.max = max;
    return this;
  }

  public getSchema(ctx?: TypeExportContext) {
    const schema: schema.ArrSchema<
      T extends void ? schema.Schema : SchemaOf<T extends Type ? T : never>,
      {[K in keyof Head]: SchemaOf<Head[K]>},
      {[K in keyof Tail]: SchemaOf<Tail[K]>}
    > = {
      ...this.schema,
    };
    const {_type, _head, _tail} = this;
    if (_type) schema.type = _type.getSchema(ctx) as any;
    if (_head) schema.head = _head.map((t) => t.getSchema(ctx)) as any;
    if (_tail) schema.tail = _tail.map((t) => t.getSchema(ctx)) as any;
    return schema;
  }

  public getOptions(): schema.Optional<
    schema.ArrSchema<T extends void ? schema.Schema : SchemaOf<T extends Type ? T : never>>
  > {
    // biome-ignore lint: unused variables are intentional
    const {kind, type, ...options} = this.schema;
    return options as any;
  }

  public toString(tab: string = ''): string {
    const {_head, _type, _tail} = this;
    return (
      super.toString(tab) +
      printTree(tab, [
        _head && _head.length
          ? (tab) =>
              '[ head, ... ]' +
              printTree(
                tab,
                _head!.map((t) => (tab) => t.toString(tab)),
              )
          : null,
        _type ? (tab) => (_type ? _type.toString(tab) : '...') : null,
        _tail && _tail.length
          ? (tab) =>
              '[ ..., tail ]' +
              printTree(
                tab,
                _tail!.map((t) => (tab) => t.toString(tab)),
              )
          : null,
      ])
    );
  }
}
