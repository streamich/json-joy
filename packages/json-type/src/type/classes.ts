import {AbsType} from './classes/AbsType';
import {AliasType} from './classes/AliasType';
import {AnyType} from './classes/AnyType';
import {ArrType} from './classes/ArrType';
import {BinType} from './classes/BinType';
import {BoolType} from './classes/BoolType';
import {ConType} from './classes/ConType';
import {FnRxType, FnType} from './classes/FnType';
import {MapType} from './classes/MapType';
import {ModuleType} from './classes/ModuleType';
import {NumType} from './classes/NumType';
import {KeyOptType, KeyType, ObjType} from './classes/ObjType';
import {OrType} from './classes/OrType';
import {RefType} from './classes/RefType';
import {StrType} from './classes/StrType';

export {
  AbsType,
  AnyType,
  ConType,
  BoolType,
  NumType,
  StrType,
  BinType,
  ArrType,
  KeyType,
  KeyOptType,
  ObjType,
  MapType,
  RefType,
  OrType,
  FnType,
  FnRxType,
  AliasType,
  ModuleType,
};
