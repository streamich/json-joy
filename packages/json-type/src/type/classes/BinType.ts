import {printTree} from 'tree-dump/lib/printTree';
import * as schema from '../../schema';
import type {SchemaOf, Type} from '../types';
import {AbsType} from './AbsType';

export class BinType<T extends Type = any> extends AbsType<schema.BinSchema> {
  constructor(
    protected type: T,
    options?: schema.Optional<schema.BinSchema>,
  ) {
    super(schema.s.Binary(schema.s.any, options));
  }

  public format(format: schema.BinSchema['format']): this {
    this.schema.format = format;
    return this;
  }

  public min(min: schema.BinSchema['min']): this {
    this.schema.min = min;
    return this;
  }

  public max(max: schema.BinSchema['max']): this {
    this.schema.max = max;
    return this;
  }

  public getSchema(): schema.BinSchema<SchemaOf<T>> {
    return {
      ...this.schema,
      type: this.type.getSchema() as any,
    };
  }

  public getOptions(): schema.Optional<schema.ArrSchema<SchemaOf<T>>> {
    // biome-ignore lint: unused variables are intentional
    const {kind, type, ...options} = this.schema;
    return options as any;
  }

  public toString(tab: string = ''): string {
    return super.toString(tab) + printTree(tab, [(tab) => this.type.toString(tab)]);
  }
}
