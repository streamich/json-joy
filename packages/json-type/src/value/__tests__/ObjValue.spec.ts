import {ModuleType} from '../../type/classes/ModuleType';
import {ObjValue} from '../ObjValue';

test('can retrieve field as Value', () => {
  const system = new ModuleType();
  const {t} = system;
  const obj = new ObjValue({foo: 'bar'}, t.Object(t.Key('foo', t.str)));
  const foo = obj.get('foo');
  expect(foo.type!.kind()).toBe('str');
  expect(foo.data).toBe('bar');
});

test('can print to string', () => {
  const system = new ModuleType();
  const {t} = system;
  const obj = new ObjValue({foo: 'bar'}, t.Object(t.Key('foo', t.str)));
  expect(obj + '').toMatchSnapshot();
});

describe('.set()', () => {
  test('can set a field', () => {
    const base = ObjValue.new();
    const t = base.system.t;
    const procedure = () => 'pong';
    const router = base.set(
      'ping',
      t.fn
        .io(t.undef, t.str)
        .ctx<{ip: string}>()
        .title('Ping the server')
        .description('Returns "pong" if the server is reachable')
        .value(procedure),
    );
    const value = router.get('ping');
    expect(value.data).toBe(procedure);
    expect(value.type!.req.getSchema()).toEqual(t.undef.getSchema());
    expect(value.type!.res.getSchema()).toEqual(t.str.getSchema());
  });

  test('can set multiple fields', () => {
    const base = ObjValue.new();
    const t = base.system.t;
    const procedure = () => 'pong';
    const router = base
      .set(
        'ping',
        t.fn
          .io(t.undef, t.str)
          .ctx<{ip: string}>()
          .title('Ping the server')
          .description('Returns "pong" if the server is reachable')
          .value(procedure),
      )
      .set(
        'echo',
        t.fn
          .io(t.any, t.any)
          .ctx<{ip: string}>()
          .title('Echo the input')
          .description('Returns the input value unchanged')
          .value((input) => input),
      )
      .set(
        'getUser',
        t.fn
          .input(t.str.title('User ID').description('ID of the user to retrieve'))
          .output(
            t.object({
              id: t.str,
              name: t.str.title('User full name').min(1).max(32),
              friends: t.fn.input(t.str).out(t.str.title('Friend name')),
            }),
          )
          .ctx<{ip: string}>()
          .title('Get user by ID')
          .description('Returns user object with the specified ID')
          .value((id) => ({id, name: 'User ' + id, friends: async (friendId) => 'Friend ' + friendId})),
      );
    expect(router.get('ping').data).toBe(procedure);
    expect(router.get('getUser').type!.req.getSchema()).toEqual(
      t.str.title('User ID').description('ID of the user to retrieve').getSchema(),
    );
    expect(router.get('getUser').type!.res.getSchema()).toEqual(
      t
        .object({
          id: t.str,
          name: t.str.title('User full name').min(1).max(32),
          friends: t.fn.input(t.str).out(t.str.title('Friend name')),
        })
        .getSchema(),
    );
    expect(router.get('echo').type!.req.getSchema()).toEqual(t.any.getSchema());
  });
});
