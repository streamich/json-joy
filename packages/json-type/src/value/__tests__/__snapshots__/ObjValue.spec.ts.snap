// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can print to string 1`] = `
"ObjValue
├─ obj
│  └─ "foo"
│      └─ str
└─ {
     "foo": "bar"
   }"
`;
